packages/puppeteer-core/src/common/Frame.ts

```diff
@@ -216,6 +221,10 @@ export class Frame {
     this._id = frameId;
     this._parentId = parentFrameId;
     this.#detached = false;
+    this.#deviceRequestPromptManager = new DeviceRequestPromptManager(
+      client,
```

Comment:
for frames the CDPSession might equal the parent's CDPSession (if the iframe is in the same process and does not have a separate debugging target). I think we should probably lazily instantiate DeviceRequestPromptManager based on the isOOPIF getter and if the frame is not an OOPIF, we should use the parent's DeviceRequestPromptManager. Perhaps you have other suggestions?

Reply:
Hmm, does something like this seem good?

```ts
  /**
   * @internal
   */
  _deviceRequestPromptManager(): DeviceRequestPromptManager {
    if (this.#deviceRequestPromptManager === undefined) {
      if (this.isOOPFrame()) {
        this.#deviceRequestPromptManager = new DeviceRequestPromptManager(
          this.#client,
          this._frameManager.timeoutSettings
        );
      } else {
        this.#deviceRequestPromptManager =
          this.parentFrame()?._deviceRequestPromptManager();
      }
    }
    return this.#deviceRequestPromptManager!;
  }

  waitForDevicePrompt(
    options: WaitTimeoutOptions = {}
  ): Promise<DeviceRequestPrompt> {
    return this._deviceRequestPromptManager().waitForDevicePrompt(options);
  }
```

I don't have any particular suggestions. I may be overlooking things about OOPIF and feature-policy controlled systems like Web Bluetooth (that the CL integrates with this devtool domain).

Reply:
This would not work when an OOPIF goes back to be a local frame. We probably should be able to create a weak map to track DeviceRequestPromptManager per client (e.g., `WeakMap<CDPSession, DeviceRequestPromptManager> map` and call `map.get(this.#client)` to get the current device request manager.

Reply:
I added a change like the above snippet with your point on using a `WeakMap.` https://github.com/puppeteer/puppeteer/pull/9299/commits/f92c95e755bb0927bc0833d4054a53dcfdbc90f3

Does that seem right? I'm not sure where the `WeakMap` would be best kept. I've added it to `FrameManager` in that commit.

Reply:
Looks good to me, but I think we need to unset the reference to the CDPSession instance somewhere in the DeviceRequestPrompt when the session gets detached. Currently, it looks like there could be a memory leak and prompts will keep the detached sessions alive. Could you check that case?

Reply:
Are you meaning a potential leak with a WeakMap value referencing its key? I don't think they work that way. I'd have to think about the [spec](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-liveness) a bit to remember why that is.

Or do you mean a leak if a user of puppeteer retains a reference to DeviceRequestPrompt directly? That'd make sense.

I've added listeners on `Target.detachedFromTarget` to unset client in `DeviceRequestPrompt` and `DeviceRequestPromptManager` and `assert` calls to ensure `#client` isn't null when its used.

Is the `Target.detachedFromTarget` event enough?

Reply:
I think it should be enough, thanks for double checking.